src/app/services/theme/theme.service.ts

```diff
@@ -0,0 +1,213 @@
+import { DOCUMENT } from "@angular/common";
+import { Inject, Injectable } from "@angular/core";
+import * as Color from "color";
+
+/** List of palette options for website theming */
+export enum BawPalettes {
+  highlight = "highlight",
+  primary = "primary",
+  secondary = "secondary",
+  success = "success",
+  info = "info",
+  warning = "warning",
+  danger = "danger",
+  light = "light",
+  dark = "dark",
+}
+
+/** List of variants which exist for each palette */
+enum BawVariants {
+  base = "",
+  lighter = "-lighter",
+  lightest = "-lightest",
+  darker = "-darker",
+  darkest = "-darkest",
+}
+
+/** List of css variables associated with a variant of a palette */
+interface BawThemeVariant {
+  color: {
+    base: string;
+    hue: string;
+    saturation: string;
+    lightness: string;
+  };
+  contrast: {
+    base: string;
+    hue: string;
+    saturation: string;
+    lightness: string;
+  };
+}
+
+/** List of palette options for website theming */
+export type BawPaletteType = keyof typeof BawPalettes;
+/** List of variants which exist for each palette */
+export type BawVariantType = keyof typeof BawVariants;
+type BawThemePalettes = { [Variant in BawVariantType]: BawThemeVariant };
+type BawThemeVariables = { [Palette in BawPaletteType]: BawThemePalettes };
+/** Configuration theme settings */
+export type BawTheme = { [Palette in BawPaletteType]?: string };
+
+@Injectable({
+  providedIn: "root",
+})
+export class ThemeService {
+  /** Tracks config changes to theme */
+  private theme: BawTheme;
+  /** Tracks all theme variables */
+  private _themeVariables: BawThemeVariables;
+  /** Track modified palettes to allow us to handle resetting styles */
+  private modifiedPalettes: Set<BawPalettes> = new Set();
```

Comment:
This is required, not all base variables are tracked by the theme (ie. default values set in styles.scss). If we delete all palettes, it'll set those values to black.

Reply:
Example? If I'm wrong here I'd like to understand why.

Of the variables we could change (`--baw-{colour}-{hue|saturation|lightness}`) which are not tracked in styles.css?

Reply:
True, you are correct. I could re-generate the `--baw-{color}` using its hue, saturation, and lightness values and do the following:

```typescript
const color = `--baw-${color}`;
this.style.setProperty(color , `hsl(var(${color}-hue), var(${color}-saturation), var(${color}-lightness))`);
```

Reply:
Actually no, i was correct the first time. We are modifying the hue, saturation, and lightness values. Effectively this is what happens. The styles.scss creates a list of values:

```css
--baw-primary: blue;
--baw-secondary: grey;
```

Then, when the application starts up we set theme theme overrides:
```css
--baw-primary: lightblue;
--baw-primary: blue;
--baw-secondary: grey;
```

If we delete `--baw-primary` using `removeProperty` everything is fine because we only remove the lightblue version, and the original remains. But if we delete `--baw-secondary`, there is no backup value, and it will be left unset

Reply:
Discussed offline, but basic browser DOM tests indicate this should be possible:
![image](https://user-images.githubusercontent.com/1452968/127590186-cccceba6-2980-40eb-8971-a3d434e7ffc7.png)

There is a difference somewhere we don't yet understand